packages/ui-app/src/Params/Param/Balance.tsx

```diff
@@ -0,0 +1,192 @@
+    const delKey = 8;
+    const escapeKey = 27;
+    const enterKey = 13;
+    const plusKey = 187; // '+'
+    const tabKey = 9; // next input field
+    const aKey = 65; // select all from balance input
+    const xKey = 88; // cut balance
+    const cKey = 67; // copy balance
+    const eKey = 69; // scientific or exponential notation
+    const vKey = 86; // paste balance
+    const zeroKey = 48;
+    const zeroNumpadKey = 96;
+
+    const regexE = /[e]/gi;
+    const regexPlus = /[\+]/gi;
+    const regexDecimalPoint = /[\.]/gi;
+
+    // only allow user balance input to contain one instance of 'e', '+', and '.' for decimal in scientific or exponential notation
+    if (
+      (event.keyCode === eKey && event.target.value.match(regexE)) ||
+      (event.keyCode === plusKey && event.target.value.match(regexPlus)) ||
+      (event.keyCode === decimalPointKey && event.target.value.match(regexDecimalPoint))
+    ) {
+      event.preventDefault();
+      return;
+    }
+
+    // obtain the current cursor index position in the Balance input field
+    const inputCursorIndex = event.target.value.slice(0, event.target.selectionStart).length;
+
+    // allow these keys
+    if ([escapeKey, enterKey, backspaceKey, delKey, tabKey, ctrlKey, altKey, cmdKey, eKey, arrowLeftKey,
+      arrowRightKey, decimalPointKey].includes(event.keyCode)) {
+      return;
+    }
+
+    // prevent user entering the '+' symbol unless immediately after the 'e' letter (i.e. 'e+') for exponential notation
+    if (event.keyCode === plusKey && event.target.value.charAt(inputCursorIndex - 1) !== 'e') {
+      event.preventDefault();
+      return;
+    }
+
+    // prevent user entering 0 at start index of input field
+    if ((event.keyCode === zeroKey || event.keyCode === zeroNumpadKey) && inputCursorIndex === 0) {
+      event.preventDefault();
+      return;
+    }
+
+    // allow users to to use cut/copy/paste combinations, but not non-numeric letters individually
+    // allow users to use the + key for exponential notation
+    if (
+      ((event.ctrlKey || cmdKey) && event.which === aKey) || // select all
+      ((event.ctrlKey || cmdKey) && event.which === cKey) || // copy
+      ((event.ctrlKey || cmdKey) && event.which === vKey) || // paste
+      ((event.ctrlKey || cmdKey) && event.which === xKey) || // cut
+      event.shiftKey && event.which === plusKey              // '+' for exponential notation (i.e. 'e+')
+    ) {
+      return;
+    }
+
+    // prevent input of non-integer values (allow numeric including from keyboards with numpad)
+    if ((event.keyCode < 48 || event.keyCode > 57) && (event.keyCode < 96 || event.keyCode > 105)) {
+      this.setState({
+        error: t('Non-numeric values are not permitted')
+      });
+      event.preventDefault();
+      return;
+    }
+  }
+
+  onKeyUp = (event: any): void => {
+    // if user inputs the value of 'E', replace it with lowercase 'e'
+    const regexE = /[E]/gi;
```

Comment:
file-based constant.

Reply:
Sorry, I'm not sure what you mean. Would you prefer I change it to say `event.target.value.indexOf('E')`?